### ISSUE TITLE
Linux: ArgumentOutOfRangeException on multi-row insert on SQL Azure

### ISSUE THREAD

**Issue by AndriySvyryd** [dotnet/corefx#7599]
I get:
```
 ---- System.Data.SqlClient.SqlException : A transport-level error has occurred when sending the request to the server. (provider: TCP Provider, error: 35 - An internal exception was caught)
 -------- System.IO.IOException : The write operation failed, see inner exception.
 ------------ System.ArgumentOutOfRangeException : Specified argument was out of the range of valid values.
 Parameter name: size
    at System.Data.SqlClient.TdsParser.TdsExecuteRPC(_SqlRPC[] rpcArray, Int32 timeout, Boolean inSchema, TdsParserStateObject stateObj, Boolean isCommandProc, Boolean sync, TaskCompletionSource`1 completion, Int32 startRpc, Int32 startParam)
    at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, SqlDataReader ds)
    at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior, String method)
    at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, String executeMethod, IReadOnlyDictionary`2 parameterValues, Boolean openConnection, Boolean closeConnection)
    at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues, Boolean manageConnection)
    at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
 ----- Inner Stack Trace -----
    at System.Net.Security.SslStreamInternal.ProcessWrite(Byte[] buffer, Int32 offset, Int32 count, AsyncProtocolRequest asyncRequest)
    at System.Net.Security.SslStream.Write(Byte[] buffer, Int32 offset, Int32 count)
    at System.Data.SqlClient.SNI.SNITCPHandle.Send(SNIPacket packet)
 ----- Inner Stack Trace -----
    at System.Net.Sockets.NetworkStream.Write(Byte[] buffer, Int32 offset, Int32 size)
    at System.Data.SqlClient.SNI.SslOverTdsStream.Write(Byte[] buffer, Int32 offset, Int32 count)
    at System.Net.Security.SslStreamInternal.StartWriting(Byte[] buffer, Int32 offset, Int32 count, AsyncProtocolRequest asyncRequest)
```
When running
```SQL
SET NOCOUNT ON;
DECLARE @toInsert2 TABLE ([Color] nvarchar(15), [DiscontinuedDate] datetime, [ListPrice] money, [Name] nvarchar(450), [ProductCategoryID] int, [ProductModelID] int, [ProductNumber] nvarchar(25), [SellEndDate] datetime, [SellStartDate] datetime, [Size] nvarchar(5), [StandardCost] money, [ThumbNailPhoto] varbinary(max), [ThumbnailPhotoFileName] nvarchar(50), [Weight] decimal, [_Position] [int]);
INSERT INTO @toInsert2
VALUES (@p0, @p1, @p2, @p3, @p4, @p5, @p6, @p7, @p8, @p9, @p10, @p11, @p12, @p13, 0),
(@p14, @p15, @p16, @p17, @p18, @p19, @p20, @p21, @p22, @p23, @p24, @p25, @p26, @p27, 1);

DECLARE @inserted2 TABLE ([ProductID] int, [Color] nvarchar(15), [DiscontinuedDate] datetime, [ListPrice] money, [ModifiedDate] datetime, [Name] nvarchar(450), [ProductCategoryID] int, [ProductModelID] int, [ProductNumber] nvarchar(25), [SellEndDate] datetime, [SellStartDate] datetime, [Size] nvarchar(5), [StandardCost] money, [ThumbNailPhoto] varbinary(max), [ThumbnailPhotoFileName] nvarchar(50), [Weight] decimal, [rowguid] uniqueidentifier, [_Position] [int]);
MERGE [SalesLT].[Product] USING @toInsert2 AS i ON 1=0
WHEN NOT MATCHED THEN
INSERT ([Color], [DiscontinuedDate], [ListPrice], [Name], [ProductCategoryID], [ProductModelID], [ProductNumber], [SellEndDate], [SellStartDate], [Size], [StandardCost], [ThumbNailPhoto], [ThumbnailPhotoFileName], [Weight])
VALUES (i.[Color], i.[DiscontinuedDate], i.[ListPrice], i.[Name], i.[ProductCategoryID], i.[ProductModelID], i.[ProductNumber], i.[SellEndDate], i.[SellStartDate], i.[Size], i.[StandardCost], i.[ThumbNailPhoto], i.[ThumbnailPhotoFileName], i.[Weight])
OUTPUT INSERTED.[ProductID], INSERTED.[Color], INSERTED.[DiscontinuedDate], INSERTED.[ListPrice], INSERTED.[ModifiedDate], INSERTED.[Name], INSERTED.[ProductCategoryID], INSERTED.[ProductModelID], INSERTED.[ProductNumber], INSERTED.[SellEndDate], INSERTED.[SellStartDate], INSERTED.[Size], INSERTED.[StandardCost], INSERTED.[ThumbNailPhoto], INSERTED.[ThumbnailPhotoFileName], INSERTED.[Weight], INSERTED.[rowguid], i._Position
INTO @inserted2;

SELECT [ProductID], [ModifiedDate], [rowguid] FROM @inserted2
ORDER BY _Position;
```
with
```SQL
`@p0 nvarchar(max) ,@p1 datetime2(7),@p2 decimal(1,0),@p3 nvarchar(4000),@p4 int,@p5 int,@p6 nvarchar(4000),@p7 datetime2(7),@p8 datetime2(7),@p9 nvarchar(max) ,@p10 decimal(1,0),@p11 varbinary(max) ,@p12 nvarchar(max) ,@p13 decimal(4,0),@p14 nvarchar(max) ,@p15 datetime2(7),@p16 decimal(1,0),@p17 nvarchar(4000),@p18 int,@p19 int,@p20 nvarchar(4000),@p21 datetime2(7),@p22 datetime2(7),@p23 nvarchar(max) ,@p24 decimal(1,0),@p25 varbinary(max) ,@p26 nvarchar(max) ,@p27 decimal(4,0)',
@p0=NULL,@p1=NULL,@p2=0,@p3=N'1705deb0-2dff-40d1-a40e-f3ff1c42e36d',@p4=NULL,@p5=NULL,@p6=N'1705deb0-2dff-40d1-a40e-f',@p7=NULL,@p8='2016-04-08 13:03:35.3168817',@p9=NULL,@p10=0,@p11=NULL,@p12=NULL,@p13=1000,@p14=NULL,@p15=NULL,@p16=0,@p17=N'5c6dc402-ab8c-4f7f-a40d-945c0491c8b8',@p18=NULL,@p19=NULL,@p20=N'5c6dc402-ab8c-4f7f-a40d-9',@p21=NULL,@p22='2016-04-08 13:03:35.8103840',@p23=NULL,@p24=0,@p25=NULL,@p26=NULL,@p27=1000
```

On Linux CoreCLR against [adventureworks.zip](https://github.com/dotnet/corefx/files/210782/adventureworks.zip)

See aspnet/EntityFramework#4228

**Comment by joshfree** [comment]
@YoungGah @saurabh500 @leecow @Petermarcu

**Comment by joshfree** [comment]
@AndriySvyryd @divega this has been a bug in the EF repo since January 5th.  Why is this only now blocking?

**Comment by AndriySvyryd** [comment]
We believed that the bug was in the way we generated the SQL and that it was fixed on our side. Our Linux tests were blocked by unrelated issues and only today we realized that the scenario didn't work there.

**Comment by joshfree** [comment]
OK.  Let's get the issue fixed and tested in the RTM branch; and then teams can discuss if this makes the bar to backport and reset rc2 escrow.

**Comment by divega** [comment]
@joshfree Sounds good. Thanks.

**Comment by saurabh500** [comment]
SqlClient creates a `SslStream` with an inner stream `SslOverTdsStream`

When the `SNIPacket` sends data to the `SslStream` it passes it down to the inner stream.
For some cases where `SNIPacket` sends data to `SslStream` the `SslOverTdsStream` receives a buffer which is smaller than the length being passed to it. 

`SslOverTdsStream` passes the buffer and length as is to its stream resulting in the exception. 

E.g. SqlClient calls SslStream.Write(…) API with a buffer of size 8000 with a length of 8000 to be written to the stream. When the inner stream gets the Write Call, the buffer size (8073) is less than the length (8085) to be written to the stream. This is not expected and hence resulting in System.ArgumentOutOfRangeException

**Comment by bartonjs** [comment]
It looks like a debug build of System.Net.Security sees this as an assert:

https://github.com/dotnet/corefx/blob/2ff9b2a1e367a9694af6bdaf9856ea12f9ae13cd/src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs#L196

Still looking.

**Comment by bartonjs** [comment]
Hmm.  SecureChannel has some math to decide how big the answer likely is (https://github.com/dotnet/corefx/blob/2ff9b2a1e367a9694af6bdaf9856ea12f9ae13cd/src/System.Net.Security/src/System/Net/SecureChannel.cs#L1061)

It's always wrong on Unix:

```
SslStream.Write(16, 0, 16)
SslStreamInternal.StartWriting(16, 0, 16)
SslState.EncryptData(16, 0, 16)
SecureChannel::Encrypt calculated bufferSizeNeeded=89 (16 + 5 + 68)
  <== outBuffer.Length=5120, (101)
SslOverTdsStream.Write(5120, 0, 101)
SslOverTdsStream.Write is deferring
NetworkStream.Write(5120, 0, 101)
SslStream.Write(16, 0, 16)
SslStreamInternal.StartWriting(16, 0, 16)
SslState.EncryptData(16, 0, 16)
SecureChannel::Encrypt calculated bufferSizeNeeded=89 (16 + 5 + 68)
  <== outBuffer.Length=5120, (101)
SslOverTdsStream.Write(5120, 0, 101)
SslOverTdsStream.Write is deferring
NetworkStream.Write(5120, 0, 101)
NetworkStream.Write(88, 0, 88)
SslOverTdsStream.Write(341, 0, 341)
SslOverTdsStream.Write is encapsulating
NetworkStream.Write(349, 0, 349)
SslOverTdsStream.Write(182, 0, 182)
SslOverTdsStream.Write is encapsulating
NetworkStream.Write(190, 0, 190)
SslStream.Write(432, 0, 432)
SslStreamInternal.StartWriting(432, 0, 432)
SslState.EncryptData(432, 0, 432)
SecureChannel::Encrypt calculated bufferSizeNeeded=505 (432 + 5 + 68)
  <== outBuffer.Length=5120, (517)
SslOverTdsStream.Write(5120, 0, 517)
SslOverTdsStream.Write is deferring
NetworkStream.Write(5120, 0, 517)
SslStream.Write(34, 0, 34)
SslStreamInternal.StartWriting(34, 0, 34)
_sslState.EncryptData(34, 0, 34)
SslState.EncryptData(34, 0, 34)
SecureChannel::Encrypt calculated bufferSizeNeeded=107 (34 + 5 + 68)
  <== outBuffer.Length=5120, (117)
SslOverTdsStream.Write(5120, 0, 117)
SslOverTdsStream.Write is deferring
NetworkStream.Write(5120, 0, 117)
SslStream.Write(8000, 0, 8000)
SslStreamInternal.StartWriting(8000, 0, 8000)
_sslState.EncryptData(8000, 0, 8000)
SslState.EncryptData(8000, 0, 8000)
SecureChannel::Encrypt calculated bufferSizeNeeded=8073 (8000 + 5 + 68)
Input buffer of size 8073 bytes is insufficient since encryption needs 8085 bytes.
```

Presumably this means either the headerSize or trailerSize calculation is wrong; but given that it's not always disagreeing by the same amount there's clearly some bad assumptions going on in this code.

**Comment by bartonjs** [comment]
Okay, there's clearly some sort of block padding miscalculation in the managed code at play.

Using a ECDHE-RSA-AES256-SHA384 connection (which is what I'm getting from the SqlAzure test):
* Writing 1 byte was 11 short
* Writing 2 was 10 short
* Writing 3 was 9 short
* Writing 4 was 8 short
* Writing 5 was 7 short
* Writing 6 was 6 short
* Writing 7 was 5 short
* Writing 8 was 4 short
* Writing 9 was 3 short
* Writing 10 was 2 short
* Writing 11 was 1 short
* Writing 12 was perfect
* Writing 13 was over-counted by 1
* Writing 14 was over-counted by 2
* Writing 15 was over-counted by 3
* Writing 16 was 12 short
* Writing 17-32 repeated the pattern

Unix reports the header as 5 bytes, trailer as 68.
Windows reports the header as 21 bytes, trailer as 64.

This all works as long as the payload stays below ~5000 bytes, since there's a "eh, it's less than 5120, use a shared buffer" codepath.

**Comment by bartonjs** [comment]
It occurs to me that I shouldn't call this "closed" since it's not yet fixed in the rc2 branch; so reopening.

**Comment by bartonjs** [comment]
This is now merged into the RC2 branch as well (#7752).
